File: XitTests/XTTest.m
#import "XTTest.h"
#import "XTRepository.h"
#import "XTRepository+Commands.h"
#import "XTRepository+Parsing.h"
#include "XTQueueUtils.h"

@implementation XTTest

- (void)setUp
{
  [super setUp];

  // /tmp is actually a link to /private/tmp, which APIs like
  // NSTemporaryDirectory and -[NSString stringByResolvingSymlinksInPath]
  // deliberately ignore, but -[NSFileManager enumeratorAtURL] doesn't.
  self.repoPath = [@"/private" stringByAppendingPathComponent:
      [NSString stringWithFormat:@"%@testrepo", NSTemporaryDirectory()]];
  self.repository = [self createRepo:self.repoPath];

  [self addInitialRepoContent];

  NSLog(@"setUp ok");
}

- (void)tearDown
{
  [self waitForRepoQueue];

  NSFileManager *defaultManager = [NSFileManager defaultManager];
  [defaultManager removeItemAtPath:self.repoPath error:nil];
  [defaultManager removeItemAtPath:self.remoteRepoPath error:nil];

  if ([defaultManager fileExistsAtPath:self.repoPath]) {
    XCTFail(@"tearDown %@ FAIL!!", self.repoPath);
  }

  if ([defaultManager fileExistsAtPath:self.remoteRepoPath]) {
    XCTFail(@"tearDown %@ FAIL!!", self.remoteRepoPath);
  }

  NSLog(@"tearDown ok");

  [super tearDown];
}

- (NSString*)file1Name
{
  return @"file1.txt";
}

- (NSString*)file1Path
{
  return [self.repoPath stringByAppendingPathComponent:self.file1Name];
}

- (NSString*)addedName
{
  return @"added.txt";
}

- (NSString*)untrackedName
{
  return @"untracked.txt";
}

- (void)makeRemoteRepo
{
  self.remoteRepoPath =
      [NSString stringWithFormat:@"%@remotetestrepo", NSTemporaryDirectory()];
  self.remoteRepository = [self createRepo:self.remoteRepoPath];
}

- (void)addInitialRepoContent
{
  XCTAssertTrue([self commitNewTextFile:self.file1Name content:@"some text"]);
}

- (void)makeStash
{
  NSError *error = nil;

  [self writeTextToFile1:@"stashy"];
  [self writeText:@"new" toFile:self.untrackedName];
  [self writeText:@"add" toFile:self.addedName];
  [self.repository stageFile:self.addedName error:&error];
  XCTAssertNil(error);
  [self.repository saveStash:@"" includeUntracked:YES];
}

- (BOOL)commitNewTextFile:(NSString *)name content:(NSString *)content
{
  return [self commitNewTextFile:name
                         content:content
                    inRepository:self.repository];
}

- (BOOL)commitNewTextFile:(NSString *)name
                  content:(NSString *)content
             inRepository:(XTRepository *)repo
{
  NSString *basePath = repo.repoURL.path;
  NSString *filePath = [basePath stringByAppendingPathComponent:name];
  NSError *error = nil;

  [content writeToFile:filePath
            atomically:YES
              encoding:NSASCIIStringEncoding
                 error:nil];

  if (![[NSFileManager defaultManager] fileExistsAtPath:filePath])
    return NO;
  if (![repo stageFile:name error:&error])
    return NO;
  if (![repo commitWithMessage:[NSString stringWithFormat:@"new %@", name]
                         amend:NO
                   outputBlock:NULL
                         error:NULL])
    return NO;

  return YES;
}

- (XTRepository *)createRepo:(NSString *)repoName
{
  NSLog(@"[createRepo] repoName=%@", repoName);
  NSFileManager *fileManager = [NSFileManager defaultManager];

  if ([fileManager fileExistsAtPath:repoName]) {
    if (![fileManager removeItemAtPath:repoName error:nil]) {
      XCTFail(@"Couldn't make way for repository: %@", repoName);
      return nil;
    }
  }
  
  BOOL created = [fileManager createDirectoryAtPath:repoName
                        withIntermediateDirectories:YES
                                         attributes:nil
                                              error:nil];
  if (!created)
    return nil;


  NSURL *repoURL = [NSURL fileURLWithPath:repoName];

  XTRepository *repo = [[XTRepository alloc] initWithURL:repoURL];

  if (![repo initializeRepository]) {
    XCTFail(@"initializeRepository '%@' FAIL!!", repoName);
  }

  if (![fileManager
          fileExistsAtPath:[NSString stringWithFormat:@"%@/.git", repoName]]) {
    XCTFail(@"%@/.git NOT Found!!", repoName);
  }

  return repo;
}

- (void)waitForRepoQueue
{
  WaitForQueue(self.repository.queue);
  WaitForQueue(dispatch_get_main_queue());
}

- (BOOL)writeText:(NSString *)text toFile:(NSString *)path
{
  return [text writeToFile:[self.repoPath stringByAppendingPathComponent:path]
                atomically:YES
                  encoding:NSUTF8StringEncoding
                     error:nil];
}

- (BOOL)writeTextToFile1:(NSString *)text
{
  NSError *error;

  [text writeToFile:self.file1Path
         atomically:YES
           encoding:NSUTF8StringEncoding
              error:&error];
  return error == nil;
}

@end


@implementation XTFakeWinController

@end

File: XitTests/XTTest.swift
import Foundation
import XCTest
@testable import Xit

class XTTest: XCTestCase
{
  var repoPath: String!
  var remoteRepoPath: String!

  var repoController: GitRepositoryController!
  var repository, remoteRepository: XTRepository!

  var file1Path: String
  { return repoPath.appending(pathComponent: TestFileName.file1.rawValue) }
  
  static func createRepo(atPath repoPath: String) -> XTRepository?
  {
    NSLog("[createRepo] repoName=\(repoPath)")
    
    let fileManager = FileManager.default
    
    if fileManager.fileExists(atPath: repoPath) {
      do {
        try fileManager.removeItem(atPath: repoPath)
      }
      catch {
        XCTFail("Couldn't make way for repository: \(repoPath)")
        return nil
      }
    }
    
    do {
      try fileManager.createDirectory(atPath: repoPath,
                                      withIntermediateDirectories: true,
                                      attributes: nil)
    }
    catch {
      XCTFail("Couldn't create repository: \(repoPath)")
      return nil
    }
    
    let repoURL = URL(fileURLWithPath: repoPath)
    let repo: XTRepository

    do {
      repo = try .init(emptyURL: repoURL)
    }
    catch let error {
      XCTFail("could not create repository at '\(repoPath)' - \(error.localizedDescription)")
      return nil
    }
    guard fileManager.fileExists(atPath: repoPath.appending(pathComponent: ".git"))
    else {
      XCTFail(".git not found")
      return nil
    }

    return repo
  }
  
  override func setUpWithError() throws
  {
    try super.setUpWithError()

    repoPath = NSString.path(withComponents: ["private",
                                              NSTemporaryDirectory(),
                                              "testRepo"])
    repository = try XCTUnwrap(XTTest.createRepo(atPath: repoPath))
    repoController = GitRepositoryController(repository: repository)
    try addInitialRepoContent()
  }
  
  override func tearDown()
  {
    waitForRepoQueue()
    
    XCTAssertNoThrow(try retryDelete(path: repoPath))
    if let remoteRepoPath = self.remoteRepoPath {
      XCTAssertNoThrow(try retryDelete(path: remoteRepoPath))
    }
    super.tearDown()
  }
  
  func retryDelete(path: String) throws
  {
    var error: Error? = nil
    
    for _ in 1...5 {
      do {
        try FileManager.default.removeItem(atPath: path)
        return
      }
      catch let e {
        error = e
      }
    }
    try error.map { throw $0 }
  }
  
  func waitForRepoQueue()
  {
      wait(for: repository)
    }
  
  func wait(for repository: XTRepository)
  {
    repository.controller?.waitForQueue()
  }
  
  func addInitialRepoContent() throws
  {
    try execute(in: repository) {
      CommitFiles {
        Write("some text", to: .file1)
      }
    }
  }
  
  func makeRemoteRepo()
  {
    let parentPath = repoPath.deletingLastPathComponent
    
    remoteRepoPath = parentPath.appending(pathComponent: "remotetestrepo")
    remoteRepository = XTTest.createRepo(atPath: remoteRepoPath)
    XCTAssertNotNil(remoteRepository)
  }

  func assertContent(_ text: String, file: String,
                     line: UInt = #line, sourceFile: StaticString = #file)
  {
    guard let content = try? String(contentsOfFile: repoPath +/ file)
    else {
      XCTFail("can't load file", file: sourceFile, line: line)
      return
    }
    guard content == text
    else {
      XCTFail("content mismatch", file: sourceFile, line: line)
      return
    }
  }
  
  func assertContent(_ text: String, file: TestFileName,
                      line: UInt = #line, sourceFile: StaticString = #file)
  {
    assertContent(text, file: file.rawValue, line: line, sourceFile: sourceFile)
  }

  func makeStash() throws
  {
    try execute(in: repository) {
      Write("stashy", to: .file1)
      Write("new", to: .untracked)
      Write("add", to: .added)
      Stage(.added)
      SaveStash()
    }
  }
}

extension RepositoryController
{
  func waitForQueue()
  {
    queue.wait()
    WaitForQueue(DispatchQueue.main)
  }
}

extension DeltaStatus: CustomStringConvertible
{
  public var description: String
  {
    switch self {
      case .unmodified:
        return "unmodified"
      case .added:
        return "added"
      case .deleted:
        return "deleted"
      case .modified:
        return "modified"
      case .renamed:
        return "renamed"
      case .copied:
        return "copied"
      case .ignored:
        return "ignored"
      case .untracked:
        return "untracked"
      case .typeChange:
        return "typeChange"
      case .conflict:
        return "conflict"
      case .mixed:
        return "mixed"
    }
  }
}
